ocaml/typing/typecore.ml

```diff
@@ -802,76 +802,35 @@ let has_poly_constraint spat =
-  let mode = Value.disallow_right mode in
-  let mode =
-    if Locality.Const.le upper_bounds.locality Locality.Const.min
-    then Value.meet_with_regionality Regionality.Const.min mode
-    else mode
-  in
-  let mode =
-    if Linearity.Const.le upper_bounds.linearity Linearity.Const.min
-    then Value.meet_with_linearity Linearity.Const.min mode
-    else mode
-  in
-  let mode =
-    if Uniqueness.Const.le upper_bounds.uniqueness Uniqueness.Const.min
-    then Value.meet_with_uniqueness Uniqueness.Const.min mode
-    else mode
-  in
-  mode
+  let upper_bounds = Const.alloc_as_value upper_bounds in
+  Value.meet_with upper_bounds mode
 
-(* cross the monadic fragment to max, and the comonadic fragment to min *)
-(* CR layouts v2.8: use the meet-with-constant morphism when available *)
+(** Mode cross a mode whose monadic fragment is a right mode, and whose comonadic
+    fragment is a left mode. *)
 let alloc_mode_cross_to_max_min env ty { monadic; comonadic } =
   let monadic = Alloc.Monadic.disallow_left monadic in
   let comonadic = Alloc.Comonadic.disallow_right comonadic in
   if not (is_principal ty) then { monadic; comonadic } else
   let jkind = type_jkind_purely env ty in
   let upper_bounds = Jkind.get_modal_upper_bounds jkind in
-  let comonadic =
-    if Locality.Const.le upper_bounds.locality Locality.Const.min
-    then Alloc.Comonadic.meet_with_locality Locality.Const.min comonadic
-    else comonadic
-  in
-  let comonadic =
-    if Linearity.Const.le upper_bounds.linearity Linearity.Const.min
-    then Alloc.Comonadic.meet_with_linearity Linearity.Const.min comonadic
-    else comonadic
-  in
-  let monadic =
-    if Uniqueness.Const.le upper_bounds.uniqueness Uniqueness.Const.min
-    then Alloc.Monadic.join_with_uniqueness Uniqueness.Const.max monadic
-    else monadic
-  in
+  let upper_bounds = Alloc.Const.split upper_bounds in
+  let comonadic = Alloc.Comonadic.meet_with upper_bounds.comonadic comonadic in
+  let monadic = Alloc.Monadic.imply upper_bounds.monadic monadic in
   { monadic; comonadic }
 
+(** Mode cross a right mode *)
 let expect_mode_cross env ty (expected_mode : expected_mode) =
   if not (is_principal ty) then expected_mode else
   let jkind = type_jkind_purely env ty in
   let upper_bounds = Jkind.get_modal_upper_bounds jkind in
-  let mode = expected_mode.mode in
-  let mode, strictly_local =
-    if Locality.Const.le upper_bounds.locality Locality.Const.min
-    then Value.join_with_regionality Regionality.Const.max mode, false
-    else mode, expected_mode.strictly_local
-  in
-  let mode =
-    if Linearity.Const.le upper_bounds.linearity Linearity.Const.min
-    then Value.join_with_linearity Linearity.Const.max mode
-    else mode
-  in
-  let mode =
-    if Uniqueness.Const.le upper_bounds.uniqueness Uniqueness.Const.min
-    then Value.join_with_uniqueness Uniqueness.Const.max mode
-    else mode
-  in
-  { expected_mode with mode; strictly_local }
```

Comment:
We don't need to update `strictly_local`, because `strictly_local` is only read when type checking functions, and functions don't cross locality.

Reply:
I agree. Please put this comment in the code, though, where it is easier for our future selves to find. Thanks.